zivid_camera/src/zivid_camera.cpp

```diff
@@ -379,6 +366,72 @@ bool ZividCamera::capture2DServiceHandler(Capture::Request&, Capture::Response&)
   return true;
 }
 
+bool ZividCamera::captureAssistantSuggestSettingsServiceHandler(CaptureAssistantSuggestSettings::Request& req,
+                                                                CaptureAssistantSuggestSettings::Response&)
+{
+  ROS_DEBUG_STREAM(__func__ << ": Request: " << req);
+
+  serviceHandlerHandleCameraConnectionLoss();
+
+  const auto max_capture_time =
+      std::chrono::round<std::chrono::milliseconds>(std::chrono::duration<double>{ req.max_capture_time.toSec() });
+  const auto ambient_light_frequency = [&req]() {
+    switch (req.ambient_light_frequency)
+    {
+      case CaptureAssistantSuggestSettings::Request::AMBIENT_LIGHT_FREQUENCY_NONE:
+        return Zivid::CaptureAssistant::AmbientLightFrequency::none;
+      case CaptureAssistantSuggestSettings::Request::AMBIENT_LIGHT_FREQUENCY_50HZ:
+        return Zivid::CaptureAssistant::AmbientLightFrequency::hz50;
+      case CaptureAssistantSuggestSettings::Request::AMBIENT_LIGHT_FREQUENCY_60HZ:
+        return Zivid::CaptureAssistant::AmbientLightFrequency::hz60;
+    }
+    throw std::runtime_error("Unhandled AMBIENT_LIGHT_FREQUENCY value: " + std::to_string(req.ambient_light_frequency));
+  }();
+
+  Zivid::CaptureAssistant::SuggestSettingsParameters suggest_settings_parameters(max_capture_time,
+                                                                                 ambient_light_frequency);
+  ROS_INFO_STREAM("Getting suggested settings using parameters: " << suggest_settings_parameters);
+  const auto suggested_settings{ Zivid::CaptureAssistant::suggestSettings(camera_, suggest_settings_parameters) };
+
+  if (suggested_settings.empty())
+  {
+    throw std::runtime_error("The suggestSettings function returned 0 settings!");
+  }
+  if (suggested_settings.size() > capture_frame_config_dr_servers_.size())
```

Comment:
It seems a strange to me that we can get an exception here for what might be a quite normal use case (f.ex. the user setting the max number of frames to 3, and the Capture Assistant then suggesting to use 4 frames). 

I guess it will improve with the possibility to specify a maximum number of frames also for the Capture Assistant.

In the mean time, why are we not using the `n=capture_frame_config_dr_servers_.size()) ` frames with highest priority from the Capture Assistant?

Reply:
I mention in the readme that the users should not lower the max number of frames parameter from 10, which is the default. And in SDK 1.8 the API should enforce that it does not return more than 10 frames. I have not yet seen more than 9 frames even with max budget. But still the API should not return more than 10 in any situation. Maybe we need to reconsider this in the future if we have a valid use case for more than 10 frames, for example by using a lot of averaging.

I don't think we should silently ignore it here and just take the N highest priority ones. One problem is that I believe the settings from the API are currently not ordered by priority.

Reply:
I agree we should not silently ignore it, we could maybe print with `ROS_WARN(...)`. However, I would be suprised as a user if I could not use the Capture Assistant if I set the max number of frames in ROS to 5 and the Capture Assistant suggested 6 frames.

Yes, if the settings from the Capture Assistant are not ordered by priority, we cannot do this.
However, we could set the number of frames / dynamic reconfigure servers in ROS fixed to be 10? Or is there a disadvantage to this?

Reply:
It is documented in the README when changing that parameter that it should not be lower than 10 if using CA.

> We do not recommend lowering this setting, especially if you are using the
> [capture_assistant/suggest_settings](#capture_assistantsuggest_settings) service.

Therefore I think it should be exception. Otherwise someone may hit this by accident, for instance if they just copied some .launch file from someplace which happened to have a lower num_capture_frames, and then they get surprised when they get bad data.

The SDK should be fixed to ensure that they are returning settings in prioritized order. @eskaur  do we have an issue on this?

We could maybe have hard coded it to 10, however it could be that for some use case like inspection customer would want to do more than 10 frames, for example doing 20 frames with averaging is a valid use case. So I made it a launch parameter. 

It would maybe be better if this number itself was a dynamic reconfigure setting. So we at runtime can add or remove these nodes as needed. It would probably make client code a bit more tricky. They would need to handle the service possibly going down and coming up again. Not sure how well that works.

Reply:
We have just made PR where max frame limit is hardcoded to 10.
It was our earlier understanding that this is a sufficient number.

Reply:
ATM the returned order is dictated by the exposure value - small exposure values targeting highlights are in the top.

Reply:
We should agree on the maximum frame limit @apartridge @kontramind

Reply:
There is agreement that the SDK will be changed to return max 10 frames. This will work with ROS with default launch parameters, and it is documented in the README that lowering the maximum is not recommended using CA. Maybe the README can be made even clearer in this regard.